crates/hir-ty/src/infer/closure.rs

```diff
@@ -146,13 +148,78 @@ pub(crate) enum CaptureKind {
 }
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub(crate) struct CapturedItem {
+pub struct CapturedItem {
     pub(crate) place: HirPlace,
     pub(crate) kind: CaptureKind,
     pub(crate) span: MirSpan,
     pub(crate) ty: Ty,
 }
 
+impl CapturedItem {
+    pub fn display_kind(&self) -> &'static str {
+        match self.kind {
+            CaptureKind::ByRef(k) => match k {
+                BorrowKind::Shared => "immutable borrow",
+                BorrowKind::Shallow => "shallow borrow",
+                BorrowKind::Unique => "unique immutable borrow",
```

Comment:
I think we probably want to hide these (that is render them as immutable borrow/mutable borrow). Most users will have no idea what these borrow kinds are as they aren't writable in the surface syntax of the language. (Alternatively, make it configurable with the default hiding them).

Reply:
Shallow is impossible for closure captures (at least I believe so, and the code I wrote won't emit them). For the unique, what about `mutable borrow (UIB, [read more](link to docs))` or something like that? They are very very rare in edition 2021 (for example r-a repo doesn't have any) so from a noise perspective I think it doesn't matter. It answers people if they wonder about why something isn't mutable if it is captured by mutable reference, and is technically more correct, but on the other hand it will distract people attention for something that doesn't really help them.

Reply:
right they are pretty rare, then I guess keeping them as is is fine (and adding a link for them so people can read up on what they are) 👍